### ISSUE TITLE
Out of memory decomposition

### ISSUE THREAD

**Issue by to266** [hyperspy/hyperspy#764]
I've been working on parts of https://github.com/hyperspy/hyperspy/issues/592, and now got to a point where decompositions are to be implemented next.

So there are a couple of ways forward and nuances:
1. Use [`dask.array.linalg.svd`](https://github.com/blaze/dask/blob/master/dask/array/linalg.py#L308) or [`dask.array.linalg.svd_compressed`](https://github.com/blaze/dask/blob/master/dask/array/linalg.py#L231) internally. These are actually only computed lazily and can even be sliced before the lazy evaluation to (ideally) reduce the actual computation. I'm not sure it's really beneficial due to the algorithm itself (it may be, have not checked), but these lazy results can then be passed to other treatments that may optionally only use a part of it.
2. Use [`dask.array.learn`](https://github.com/blaze/dask/issues/817) directly. I suppose it can be combined with the previous point, hence enabling simple ICA and (however computationally difficult) NMF.
3. Just use [incremental PCA](http://scikit-learn.org/stable/modules/decomposition.html#incremental-pca) directly by just slicing the correct part of the data. That only covers PCA, though, and can (probably) be a separate way to read data (i.e. perform the PCA while reading the large and/or live dataset, and only leave the decomposition results).
4. `mask` in any large data decomposition will be extremely difficult or impossible to implement. The dask constraint is, in a sentence "Can infer the shape of the array at any point", and masking would change the shape depending on values, hence not really compatible. At the moment I would just disable all masks and any other unnecessary pre- and post-treatments (i.e. centering, which would require to sum the data, which is expensive on GB-sized signals)
5. The quite recent decomposition change made it a requirement for the learning results to be signals that get sliced. In my current implementation there are no "lazy signals", so every time a big signal is sliced, a temporary file is created on a disk, where the result is written. I don't think it would be too difficult to actually add lazy signals, however do we really want it? I suppose they would also be useful when plotting navigators, but would have to check if it works (again, at the moment a temp file is created that contains the "sum" spectrum, that gets plotted as the navigator). The main drawback is that the computations (when accumulated) can be take a lot of time or just crash due to not having enough resources (it may still happen, because defining dask arrays requires defining chunks to work with. Naturally the larger the chunk, the faster the performance, however some operations may require 2x or more memory than the chunk alone, hence possibility)

**Comment by dnjohnstone** [comment]
Did any further progress get made with this yet?

If it's not too late, I think option 2 sounds like the best to me, simply because, if I understand correctly it's the only one that may enable ICA & NMF... I really think that being able to do those two on heavy datasets is going to be extremely valuable for many microscopists at least.

**Comment by tjof2** [comment]
Does option 2 involve the appropriate whitening/denoising methods with PCA? I've not read the code in detail, but as @to266 says, it would likely need combining with option 1.

Separately, @dnjohnstone, I've returned to tensor methods quickly, the fast non-negative tensor decomposition we've used on your data might be better implemented directly in Python using the relevant code in this [scikit-tensor file](https://github.com/mnick/scikit-tensor/blob/master/sktensor/tucker.py). Separate-but-related issue worth discussing?
